tests/table/test_locations.py

```diff
@@ -0,0 +1,130 @@
+PARTITION_SPEC = PartitionSpec(PARTITION_FIELD)
+PARTITION_KEY = PartitionKey(
+    raw_partition_field_values=[PartitionFieldValue(PARTITION_FIELD, "example#val")],
+    partition_spec=PARTITION_SPEC,
+    schema=TABLE_SCHEMA,
+)
+
+
+class CustomLocationProvider(LocationProvider):
+    def new_data_location(self, data_file_name: str, partition_key: Optional[PartitionKey] = None) -> str:
+        return f"custom_location_provider/{data_file_name}"
+
+
+def test_default_location_provider() -> None:
+    provider = load_location_provider(table_location="table_location", table_properties=EMPTY_DICT)
+
+    assert provider.new_data_location("my_file") == "table_location/data/my_file"
+
+
+def test_custom_location_provider() -> None:
+    qualified_name = CustomLocationProvider.__module__ + "." + CustomLocationProvider.__name__
+    provider = load_location_provider(
+        table_location="table_location", table_properties={"write.location-provider.impl": qualified_name}
+    )
+
+    assert provider.new_data_location("my_file") == "custom_location_provider/my_file"
+
+
+def test_custom_location_provider_single_path() -> None:
+    with pytest.raises(ValueError, match=r"write\.location-provider\.impl should be full path"):
+        load_location_provider(table_location="table_location", table_properties={"write.location-provider.impl": "not_found"})
+
+
+def test_custom_location_provider_not_found() -> None:
+    with pytest.raises(ValueError, match=r"Could not initialize LocationProvider"):
+        load_location_provider(
+            table_location="table_location", table_properties={"write.location-provider.impl": "module.not_found"}
+        )
+
+
+def test_object_storage_injects_entropy() -> None:
+    provider = load_location_provider(table_location="table_location", table_properties={"write.object-storage.enabled": "true"})
+
+    location = provider.new_data_location("test.parquet")
+    parts = location.split("/")
+
+    assert len(parts) == 7
+    assert parts[0] == "table_location"
+    assert parts[1] == "data"
+    # Entropy directories in the middle
+    assert parts[-1] == "test.parquet"
+
+
+@pytest.mark.parametrize("object_storage", [True, False])
+def test_partition_value_in_path(object_storage: bool) -> None:
+    provider = load_location_provider(
+        table_location="table_location",
+        table_properties={
+            "write.object-storage.enabled": str(object_storage),
+        },
+    )
+
+    location = provider.new_data_location("test.parquet", PARTITION_KEY)
+    partition_segment = location.split("/")[-2]
+
+    # Field name is not encoded but partition value is - this differs from the Java implementation
+    # https://github.com/apache/iceberg/blob/cdf748e8e5537f13d861aa4c617a51f3e11dc97c/core/src/test/java/org/apache/iceberg/TestLocationProvider.java#L304
+    assert partition_segment == "part#field=example%23val"
```

Comment:
Thinking of opening up a separate issue/PR about this (it seems odd that partition values are encoded but field names aren't). In this case, I can remove the special-character testing (that the Java counterpart does) because it'll be unit-tested in that PR.